src/psyclone/domain/lfric/lfric_stencils.py

```diff
@@ -0,0 +1,649 @@
+    (PSy, Invokes, Invoke, InvokeSchedule, Loop, Kern, Inf, Arguments and
+    Argument). '''
+
+from psyclone.configuration import Config
+from psyclone.domain.lfric.lfric_collection import LFRicCollection
+from psyclone.domain.lfric.lfric_constants import LFRicConstants
+from psyclone.errors import GenerationError, InternalError
+from psyclone.f2pygen import (AssignGen, CommentGen, DeclGen,
+                              IfThenGen, TypeDeclGen, UseGen)
+from psyclone.psyir.symbols import ScalarType
+
+class LFRicStencils(LFRicCollection):
+    '''
+    Stencil information and code generation associated with a PSy-layer
+    routine or Kernel stub.
+
+    :param node: the Invoke or Kernel stub for which to provide stencil info.
+    :type node: :py:class:`psyclone.dynamo0p3.LFRicInvoke` or \
+                :py:class:`psyclone.domain.lfric.LFRicKern`
+
+    :raises GenerationError: if a literal has been supplied for a stencil \
+                             direction.
+    '''
+    def __init__(self, node):
+        # pylint: disable=too-many-branches
+        super().__init__(node)
+
+        # List of arguments which have an extent value passed to this
+        # invoke routine from the algorithm layer. Duplicate argument
+        # names are removed.
+        self._unique_extent_args = []
+        extent_names = []
+        # pylint: disable=too-many-nested-blocks
+        for call in self._calls:
+            for arg in call.arguments.args:
+                if arg.stencil:
+                    # Check for the existence of arg.extent here as in
+                    # the future we plan to support kernels which
+                    # specify the value of extent in metadata. If this
+                    # is the case then an extent argument is not
+                    # required.
+                    # TODO #963
+                    if not arg.stencil.extent:
+                        if not arg.stencil.extent_arg.is_literal():
+                            if arg.stencil.extent_arg.text not in extent_names:
+                                extent_names.append(
+                                    arg.stencil.extent_arg.text)
+                                self._unique_extent_args.append(arg)
+
+        # A list of arguments that have a direction variable passed in
+        # to this invoke routine from the algorithm layer. Duplicate
+        # argument names are removed.
+        self._unique_direction_args = []
+        direction_names = []
+        for call in self._calls:
+            for idx, arg in enumerate(call.arguments.args):
+                if arg.stencil and arg.stencil.direction_arg:
+                    if arg.stencil.direction_arg.is_literal():
+                        raise GenerationError(
+                            f"Kernel {call.name}, metadata arg {idx}, a "
+                            f"literal is not a valid value for a stencil "
+                            f"direction")
```

Comment:
I would suggest adding a full stop in this error message. The related test will need to be adjusted, too.